### PR TITLE
homework_2: Create REST-API to handle users

### PR DESCRIPTION
**Pull request by Roma-Ol**

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+dotenv.config();
+
+const port = process.env.PORT || 8080;
+
+app.use(express.json());
+app.use(morgan('dev'));
+app.use('/', usersRouter);
+
+// Centralized requests error handler.
+app.use((err, req, res, next) => {
```
Comment (YuriiShendryk) [inline server.js]: this default error handler should be added at the end of the middleware function stack, in your case after the bad request error handler

File: server.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  res.status(err.statusCode || 500).json({
+    status: 'error',
+    message: err.message || 'Internal Server Error'
+  });
+});
+
+// 404 request Handler.
+app.use((req, res, next) => {
+  res.status(404).json({
```
Comment (YuriiShendryk) [inline server.js]: It would be good to create a variable to store 400 status code.

File: server.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    status: 'error',
+    code: 404,
+    message: 'Not Found'
```
Comment (YuriiShendryk) [inline server.js]: create a separate variable to store 400 status code, please

File: src/controllers/index.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  getUserById,
+  createUser,
+  updateUser,
+  deleteUser
+} = require('../services');
+
+const getAllUsersHandler = async (req, res, next) => {
+  try {
+    const users = await getAllUsers();
+    res.status(200).json(users);
```
Comment (YuriiShendryk) [inline src/controllers/index.js]: The sane as for 400 status code. You can create the statusCodes object where you can store different statuses.

File: src/controllers/index.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  } catch (err) {
+    next(err);
+  }
+};
+
+const getUserByIdHandler = async (req, res, next) => {
+  try {
+    const {id} = req.params;
+    const selectedUser = await getUserById(id);
+    res.status(200).json(selectedUser);
```
Comment (YuriiShendryk) [inline src/controllers/index.js]: Create variable for the 200 status code to avoid magic numbers

File: src/controllers/index.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  } catch (err) {
+    next(err);
+  }
+};
+
+const createUserHandler = async (req, res, next) => {
+  try {
+    const newUser = await createUser(req.body);
+
+    res.status(201).json(newUser);
```
Comment (YuriiShendryk) [inline src/controllers/index.js]: Separate variable

File: src/controllers/index.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    next(err);
+  }
+};
+
+const deleteUserByIdHandler = async (req, res, next) => {
+  try {
+    const {id} = req.params;
+    await deleteUser(id);
+
+    res.status(204).send();
```
Comment (YuriiShendryk) [inline src/controllers/index.js]: Separate variable

File: src/controllers/index.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    next(err);
+  }
+};
+
+const updateUserHandler = async (req, res, next) => {
+  try {
+    const {id} = req.params;
+    const updatedUser = await updateUser(id, req.body);
+
+    res.status(200).json(updatedUser);
```
Comment (YuriiShendryk) [inline src/controllers/index.js]: Separate variable

File: src/middlewares/userValidate.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Validation middleware.
+ */
+const Joi = require('joi');
+
+const createUserSchema = Joi.object({
+  name: Joi.string().required().min(3).max(30).message('Name must be between 3 and 30 characters'),
+  favorite_color: Joi.string().required().max(20).message('Favorite color must be 20 characters or less'),
+  age: Joi.number().integer().required().min(0).max(120).message('Age must be a valid integer between 0 and 120'),
```
Comment (YuriiShendryk) [inline src/middlewares/userValidate.js]: I think it would be better to make min age equals to 16 e.g, but not to 0

File: server.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+const port = process.env.PORT || 8080;
+
+app.use(express.json());
+app.use(morgan('dev'));
+app.use('/', usersRouter);
+
+// Centralized requests error handler.
+app.use((err, req, res, next) => {
+  res.status(err.statusCode || 500).json({
```
Comment (YuriiShendryk) [inline server.js]: Move 500 into separate variable

File: src/services/index.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+const userStoragePath = pathJoin(__dirname, '../models/users.json');
+
+const verifyUserExists = async (userId) => {
+  const allUsers = await getAllUsers();
+  const user = allUsers.find(({ id }) => id === userId);
+
+  if (!user) {
+    const error = new Error(`User with ID ${userId} not found`);
+    error.statusCode = 404;
```
Comment (YuriiShendryk) [inline src/services/index.js]: Use separate variable instead of the 404 status code

File: src/services/index.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+ * returns the updated entity from controller.
+ */
+const { readFile, writeFile } = require('fs/promises');
+const { join: pathJoin } = require('path');
+const { updateUserStorage } = require('../utils/updateUserStorage');
+const { v4: uuidv4 } = require('uuid');
+
+const userStoragePath = pathJoin(__dirname, '../models/users.json');
+
+const verifyUserExists = async (userId) => {
```
Comment (YuriiShendryk) [inline src/services/index.js:12]: You can return user from this function, but I think, in this case you have to rename it

File: src/services/index.js [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+};
+
+const getAllUsers = async () => {
+  const allUsers = await readFile(userStoragePath, 'utf8');
+
+  return JSON.parse(allUsers);
+};
+
+const getUserById = async (userId) => {
+  await verifyUserExists(userId);
```
Comment (YuriiShendryk) [inline src/services/index.js:30]: If you return user from this function, you don't have to call getAllUsers() again in next code lines

File: src/services/index.js [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+  const allUsers = await getAllUsers();
+  return allUsers.find(({ id }) => id === userId);
+};
+
+const createUser = async (userData) => {
+  const allUsers = await getAllUsers();
+  const newUser = { ...userData, id: uuidv4() };
+  await updateUserStorage([...allUsers, newUser]);
```
Comment (YuriiShendryk) [inline src/services/index.js:38]: In this case it is Ok to move this logic in the helpers forlder, but I think that this functionality related to users only. Therefore you can keep this functionality in this service.

File: src/models/users.json [hunk 15]
```diff
@@ -0,0 +1 @@
+[{"name":"Santiago","age":"11","favorite_color":"blue","id":"1ffa27ab-a641-46af-8a75-acc06dbf571d"},{"name":"Pedro","age":"43","favorite_color":"pink","id":"1ffa27ab-a641-3333-8a75-acc06dbf571d"},{"name":"Lisa","age":"34","favorite_color":"red","id":"31ce22ab-4e11-4ec5-b457-23855d896236"}]
```
Comment (YuriiShendryk) [inline src/models/users.json]: user have to consists of name, email, phone, age. I don't see an email and phone

File: src/middlewares/userValidate.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+});
+
+const updateUserSchema = Joi.object({
+  name: Joi.string().min(3).max(30).message('Name must be between 3 and 30 characters'),
+  favorite_color: Joi.string().max(20).message('Favorite color must be 20 characters or less'),
+  age: Joi.number().integer().min(0).max(120).message('Age must be a valid integer between 0 and 120'),
```
Comment (YuriiShendryk) [inline src/middlewares/userValidate.js]: user have to consists of name, email, phone, age. I don't see an email and phone

**Review by YuriiShendryk — COMMENTED** [review]
Good job

File: src/services/index.js (hunk 13 above)
Comment (Roma-Ol) [reply src/services/index.js:30]: I'll do that in the next PR not to create conflicts

**Pull request merged**